pkg/nodeagent/controller/operatingsystemconfig/reconciler.go

```diff
@@ -44,17 +50,30 @@ import (
 	filespkg "github.com/gardener/gardener/pkg/nodeagent/files"
 	"github.com/gardener/gardener/pkg/nodeagent/registry"
 	"github.com/gardener/gardener/pkg/utils/flow"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+	retryutils "github.com/gardener/gardener/pkg/utils/retry"
 )
 
 const (
 	lastAppliedOperatingSystemConfigFilePath         = nodeagentconfigv1alpha1.BaseDir + "/last-applied-osc.yaml"
 	lastComputedOperatingSystemConfigChangesFilePath = nodeagentconfigv1alpha1.BaseDir + "/last-computed-osc-changes.yaml"
+	annotationUpdatingOSVersion                      = "node-agent.gardener.cloud/updating-os-version"
 )
 
-var codec runtime.Codec
+var (
+	codec                         runtimepkg.Codec
+	osVersionRegex                = regexp.MustCompile(`\d+(?:\.\d+)+`)
+	retriableErrorPatternRegex    = regexp.MustCompile(`(?i)network problems`)
+	nonRetriableErrorPatternRegex = regexp.MustCompile(`(?i)invalid arguments|system failure`)
```

Comment:
I have just added this for the completeness of this PR. We can refine it later according to gardenlinux. 
/cc @nkraetzschmar